source/unifiedbuffer.ts

```diff
@@ -0,0 +1,142 @@
+
+/* spellchecker: disable */
+
+import { Buffer } from './buffer';
+import { Context } from './context';
+import { Initializable } from './initializable';
+
+/* spellchecker: enable */
+
+
+export class UnifiedBuffer extends Initializable {
+
+    protected _cpuBuffer: ArrayBuffer;
+    protected _gpuBuffer: Buffer;
+    protected _updates = new Array<Update>();
+
+    usage: GLenum;
+
+    constructor(context: Context, sizeInBytes: number, usage: GLenum, identifier?: string) {
+        super();
+
+        this._cpuBuffer = new ArrayBuffer(sizeInBytes);
+        this._gpuBuffer = new Buffer(context, identifier);
+        this.usage = usage;
+    }
+
+    protected _addUpdate(update: Update): void {
+        const toRemove = new Array<number>();
+        const toMerge = new Array<Update>();
+
+        // Mark all older updates which overlap with the new update
+        // for removing / merging
+        this._updates.forEach((current: Update, index: number) => {
+            if (!(update.end < current.begin || update.begin > current.end)) {
+                toRemove.push(index);
+                toMerge.push(current);
+            }
+        });
+
+        // Remove all older updates, since they get merged in the next step
+        for (let i = toRemove.length - 1; i >= 0; i--) {
+            this._updates.splice(toRemove[i], 1);
+        }
+
+        // Finally merge all overlapping updates
+        toMerge.push(update);
+        const begin = Math.min(...toMerge.map((merge: Update) => merge.begin));
+        const end = Math.max(...toMerge.map((merge: Update) => merge.end));
+
+        this._updates.push({ begin, end });
+    }
+
+    @Initializable.initialize()
+    initialize(target: GLenum): boolean {
+        return this._gpuBuffer.initialize(target);
```

Comment:
Initialization (and re-initialization) requires the whole CPU-buffer to be retransferred to the GPU.
I suggest to either push the change here or in the `uninitialize` function.

Reply:
Currently this happens lazily when update is called. I did this so all "heavy" operations (allocation, transfer) only take place in update.
Do you want both the allocation and transfer on re-initialization to happen in initialize / uninitialize or only the allocation?
And if the allocation (+ transfer) on re-initialization happens in initialize wouldn't it be more consistent to do the same for resize instead of doing in update?

Reply:
Okay, I missed this behavior. This could use an additional comment:
´´´
// This full initialization of the buffer is triggered after resize, recreation, and context-lost.
´´´